### PR TITLE
add module for yak

### PR DESCRIPTION
**Pull request by yumisims**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/yak/count/meta.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+input:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
+  - fastqs:
+      type: path
+      description: directory contains compressed reads fastq file
+      pattern: "*.{fastq.gz,fq.gz}"
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/meta.yml]: Can also be fasta.

File: modules/nf-core/yak/count/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        when {
+            process {
+                """                
+                input[0] = [
+                    [ id:'test' ], // meta map
+                    file(params.test_data['sarscov2']['illumina']['test_methylated_1_fastq_gz'], checkIfExists: true)
+                ]
+                """
+            }
+        }
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/tests/main.nf.test:23]: I don't think you need the methylated files right, if not don't overcomplicate it. 

```suggestion
    test("sarscov2 - Illumina - se") {

        when {
            process {
                """                
                input[0] = [
                    [ id:'test' ], // meta map
                   file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true),
                ]
                """
            }
        }
```

File: modules/nf-core/yak/count/main.nf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    output:
+    tuple val(meta), path("*.yak"), emit: yak
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/main.nf:22]: To accomodate for pe think it needs to be something like this. Also add a test for it
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
    input_command = meta.single_end ? "${fastq}" : "<(zcat ${fastq}) <(zcat ${fastq})"
```

File: modules/nf-core/yak/count/main.nf [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    yak \\
+        count \\
+        $args \\
+        -t${task.cpus} \\
+        -o ${prefix}.yak\\
+        $fastq
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/main.nf:29]:
```suggestion
        $input_command
```

File: modules/nf-core/yak/count/main.nf [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        yak: \$(yak --version |& sed '1!d ; s/yak //')
+    END_VERSIONS
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/main.nf:39]: add input_command

File: modules/nf-core/yak/count/tests/main.nf.test [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
+            )
+        }
+
+    }
+
+    // TODO nf-core: Change the test name preferably indicating the test-data and file-format used but keep the " - stub" suffix.
+    test("test_yak_count_methylated_1_fastq_gz - stub") {
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/tests/main.nf.test:35]: - remove todo statement
- add pe test
- update test name

**Review by Joon-Klaps — CHANGES_REQUESTED** [review]
Looks great! Minor suggestions.

File: modules/nf-core/yak/count/meta.yml (hunk 1 above)
Comment (yumisims) [reply modules/nf-core/yak/count/meta.yml]: thanks, I have added fasta option as you advised.

File: modules/nf-core/yak/count/tests/main.nf.test (hunk 6 above)
Comment (yumisims) [reply modules/nf-core/yak/count/tests/main.nf.test:35]: you are right, I was thinking of having a step to concatenate the two fastqs before the yak, amazing solution.

I will incorporate your suggestion and resubmit :)

Thank you

File: modules/nf-core/yak/count/main.nf [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    input_command = meta.single_end ? "${fastq}" : "<(zcat ${fastq}) <(zcat ${fastq})"
+    """
+    yak \\
+        count \\
+        $args \\
+        -t${task.cpus} \\
+        -o ${prefix}.yak\\
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/main.nf:29]:
```suggestion
        -o ${prefix}.yak \\
```

File: modules/nf-core/yak/count/meta.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
+name: "yak_count"
+description: a tool to build k-mer hash table for assembly; count singletons
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/meta.yml:2]:
```suggestion
description: a tool to build k-mer hash table for fasta and fastq files 
```

File: modules/nf-core/yak/count/meta.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+keywords:
+  - k-mers
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/meta.yml:4]:
```suggestion
  - kmer
  - fastq
  - sequence
```

File: modules/nf-core/yak/count/tests/main.nf.test [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    test("sarscov2 - Illumina - pe") {
+
+        when {
+            process {
+                """                
+                input[0] = [
+                    [ id:'test', single_end:false ], // meta map
+                    file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true)
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/tests/main.nf.test:65]: Think you forgot to add the other read pair
```suggestion
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_1.fastq.gz", checkIfExists: true),
                     file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/fastq/test_2.fastq.gz", checkIfExists: true)]
                    
```

File: modules/nf-core/yak/count/tests/main.nf.test [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+        then {
+            assertAll(
+                { assert process.success },
+                { assert snapshot(process.out).match() }
+            )
+        }
+
+    }
+
+    test("sarscov2 - Illumina - se - stub") {
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/tests/main.nf.test:34]: If possible move so it's the final test

File: tests/modules/nf-core/yak/count/main.nf
Comment (Joon-Klaps) [inline tests/modules/nf-core/yak/count/main.nf:1]: Sorry, this totally slipped me previously. This testing method has been replaced with the nf-test framework, we don't use them anymore.  
So this fille should be deleted

File: tests/modules/nf-core/yak/count/nextflow.config
Comment (Joon-Klaps) [inline tests/modules/nf-core/yak/count/nextflow.config:1]: can remove

File: tests/modules/nf-core/yak/count/test.yml
Comment (Joon-Klaps) [inline tests/modules/nf-core/yak/count/test.yml:1]: can remove

File: modules/nf-core/yak/count/meta.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'sample1', single_end:false ]`
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - yak:
+      type: file
+      description: k-mer hash table for assembly
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/meta.yml:39]:
```suggestion
      description: k-mer hash table of input
```

File: modules/nf-core/yak/count/main.nf [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        $input_command
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        yak: \$(yak --version |& sed '1!d ; s/yak //')
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/main.nf:34]:
```suggestion
        yak: \$(yak version)
```
Think u missed this, as the output was `yak: [E::main] unknown command`

File: modules/nf-core/yak/count/main.nf [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    touch ${prefix}.yak
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        yak: \$(yak version |& sed '1!d ; s/yak //')
```
Comment (Joon-Klaps) [inline modules/nf-core/yak/count/main.nf:46]:
```suggestion
        yak: \$(yak version)
```

**Review by Joon-Klaps — CHANGES_REQUESTED** [review]
Nice, a couple of small suggestions again. Remake snapshot and we are good to go!

File: modules/nf-core/yak/count/tests/main.nf.test (hunk 11 above)
Comment (yumisims) [reply modules/nf-core/yak/count/tests/main.nf.test:34]: @Joon-Klaps 
Hi Joon,

Thanks for your very constructive comments and reviews. I might have misunderstood, but did you mean to remove the pe test?

Thanks
Yumi

File: tests/modules/nf-core/yak/count/main.nf
Comment (yumisims) [reply tests/modules/nf-core/yak/count/main.nf:1]: sorry, which file? tests/modules/nf-core/yak/count/main.nf?

File: modules/nf-core/yak/count/tests/main.nf.test (hunk 11 above)
Comment (Joon-Klaps) [reply modules/nf-core/yak/count/tests/main.nf.test:34]: no sorry for being unclear, I meant to move the stub test as a final test in the nf.test file. it's more common to have the stub test just as the last test. So it's 1: se, 2:pe, 3:stub.

File: tests/modules/nf-core/yak/count/nextflow.config
Comment (Joon-Klaps) [inline tests/modules/nf-core/yak/count/nextflow.config:1]: I meant delete delete, these files. They are redundant now. 
We only need test files that are in `modules/nf-core/yak/count/tests/`. The other test files are no longer needed.

File: modules/nf-core/yak/count/tests/main.nf.test (hunk 11 above)
Comment (yumisims) [reply modules/nf-core/yak/count/tests/main.nf.test]: thanks Joon, all done now.

File: tests/modules/nf-core/yak/count/nextflow.config
Comment (yumisims) [reply tests/modules/nf-core/yak/count/nextflow.config:1]: good to know that test is no longer needed. Thanks Joon

**Review by Joon-Klaps — APPROVED** [review]
Awesome! 🚀

**Pull request merged**